doc/tutorials/klondike/app/lib/step5/klondike_game.dart

```diff
@@ -0,0 +1,294 @@
+import 'dart:ui';
+
+import 'package:flame/components.dart';
+import 'package:flame/flame.dart';
+import 'package:flame/game.dart';
+
+import 'components/card.dart';
+import 'components/flat_button.dart';
+import 'components/foundation_pile.dart';
+import 'components/stock_pile.dart';
+import 'components/tableau_pile.dart';
+import 'components/waste_pile.dart';
+
+enum Startup { first, newDeal, sameDeal, changeDraw, haveFun }
+
+class KlondikeGame extends FlameGame {
+  static const double cardGap = 175.0;
+  static const double topGap = 500.0;
+  static const double cardWidth = 1000.0;
+  static const double cardHeight = 1400.0;
+  static const double cardRadius = 100.0;
+  static const double cardSpaceWidth = cardWidth + cardGap;
+  static const double cardSpaceHeight = cardHeight + cardGap;
+  static final Vector2 cardSize = Vector2(cardWidth, cardHeight);
+  static final cardRRect = RRect.fromRectAndRadius(
+    const Rect.fromLTWH(0, 0, cardWidth, cardHeight),
+    const Radius.circular(cardRadius),
+  );
+
+  // Default is Klondike Draw 1: a button switches between Draw 1 and Draw 3.
+  var _klondikeDraw = 1;
+
+  int get klondikeDraw => _klondikeDraw;
+
+  final stock = StockPile(position: Vector2(cardGap, topGap));
+  final waste = WastePile(position: Vector2(cardSpaceWidth + cardGap, topGap));
+  final List<FoundationPile> foundations = [];
+  final List<TableauPile> piles = [];
+  final List<Card> cards = [];
+
+  @override
+  Future<void> onLoad() async {
+    await Flame.images.load('klondike-sprites.png');
+
+    for (var i = 0; i < 4; i++) {
+      foundations.add(
+        FoundationPile(
+          i,
+          checkWin,
+          position: Vector2((i + 3) * cardSpaceWidth + cardGap, topGap),
+        ),
+      );
+    }
+    for (var i = 0; i < 7; i++) {
+      piles.add(
+        TableauPile(
+          position: Vector2(
+            i * cardSpaceWidth + cardGap,
+            cardSpaceHeight + topGap,
+          ),
+        ),
+      );
+    }
+
+    for (var rank = 1; rank <= 13; rank++) {
+      for (var suit = 0; suit < 4; suit++) {
+        cards.add(Card(rank, suit));
+      }
+    }
+
+    final gameSize =
+        Vector2(7 * cardSpaceWidth + cardGap, 4 * cardSpaceHeight + topGap);
+    final gameMidX = gameSize.x / 2;
+
+    world.add(stock);
+    world.add(waste);
+    world.addAll(foundations);
+    world.addAll(piles);
+    world.addAll(cards);
+
+    addButton('New deal', gameMidX, Startup.newDeal);
+    addButton('Same deal', gameMidX + cardSpaceWidth, Startup.sameDeal);
+    addButton('Draw 1 ⇌ 3', gameMidX + 2 * cardSpaceWidth, Startup.changeDraw);
+    addButton('Have fun', gameMidX + 3 * cardSpaceWidth, Startup.haveFun);
+
+    camera.viewfinder.visibleGameSize = gameSize;
+    camera.viewfinder.position = Vector2(gameMidX, 0);
+    camera.viewfinder.anchor = Anchor.topCenter;
+
+    init(Startup.first);
+  }
+
+  void init(Startup startType) {
```

Comment:
All the `init` methods are a bit strange, and error prone.
It would be better if all components are completely re-created when a game is reset, instead of being modified (since it is much easier for the user to forget to reset a bit of state then).
To re-play with the deck in a certain order the previos `Random` seed can be passed to shuffle and it'll put them in the same order..

Reply:
I'm still thinking about this.

For me, I think it would be more time-consuming and error-prone to try to re-create all components than to get initializations right in five classes which only require setting a few private state-variables back to their intial values. I have had quite a bit of experience with this pre-Flutter and pre-Dart.

OTOH I have no idea how to find all the existing components in the `World`, nor get them out and replace them with others, nor create a suitable seed to pass in for each new deal, nor even how to pass a seed to `Random`. Why the square brackets in the API description?

Maybe I should try the approach of having a `restart()` callback to replace the GameWidget after extracting a few key data-items to pass back in after the restart.

Reply:
> For me, I think it would be more time-consuming and error-prone to try to re-create all components than to get initializations right in five classes which only require setting a few private state-variables back to their intial values. I have had quite a bit of experience with this pre-Flutter and pre-Dart.

Why do you think that would be more error prone? Like it is right now you will have two different code paths of getting to a fully initialized component, either a completely fresh component like you will get on the first round that you're playing, or the second path would be when you reset components that have already existed in a previous game. I would say that in all type of coding you want to minimize the amount of code paths that end up in the same state, the only exception is if you do it because of performance reasons, but with as few and lightweight components as it is in here that is not a concern.

> OTOH I have no idea how to find all the existing components in the World, nor get them out and replace them with others, nor create a suitable seed to pass in for each new deal, nor even how to pass a seed to Random. Why the square brackets in the API description?

You don't have to find them, instead of adding them to the world from the game's `onLoad` create a whole new world component, and inside of that world component's `onLoad` you do pretty much everything that is done in the game's `onLoad` now.
Something like this:

```dart
class KlondikeWorld extends World {
  KlondikeWorld({this.klondikeDraw = 1, this.seed});
  
  // Default is Klondike Draw 1: a button switches between Draw 1 and Draw 3.
  final int klondikeDraw;
  // The seed used to make the randomization deterministic.
  int? seed;
  late Random random;

  final stock = StockPile(position: Vector2(cardGap, topGap));
  final waste = WastePile(position: Vector2(cardSpaceWidth + cardGap, topGap));
  final List<FoundationPile> foundations = [];
  final List<TableauPile> piles = [];
  final List<Card> cards = [];

  @override
  Future<void> onLoad() async {
    // This creates a random seed if one was not supplied.
    seed ??= Random().nextInt();
    // This creates the random generator that should be used everywhere for shuffling etc.
    random = Random(seed);
    await Flame.images.load('klondike-sprites.png');
    ...
    // This would maybe not be done in this method, just showing how to pass the random class to it.
    cards.shuffle(random);
   }
}
```

And then when you restart the game you can simply do:
```dart
game.world = World(klondikeDraw: 3);
```
And if you then would like to play the same setup as you had before:
```dart
game.world = World(
  klondikeDraw: game.world.klondikeDraw, 
  seed: game.random.seed,
);
```

> Why the square brackets in the API description?

The square brackets mean that it is an optional positional argument (not named), so that you just call it like this:
```dart
// Setting 5 as seed as an example.
final random = Random(5);
```

> Maybe I should try the approach of having a restart() callback to replace the GameWidget after extracting a few key data-items to pass back in after the restart.

You could also do that! That is a completely legit way of doing it too, but you'd have to get the seed to the state somehow then.

Apologize for the long comment, since this tutorial will be our prime tutorial of the documentation after your renovation of it I think it would be great to have everything top notch. And also, thanks for putting time into this!

Reply:
Thank you, Spydon, for your prompt reply to my earlier concerns and for your very kind final paragraph. It is I who should apologise, and I do, for taking up so much of your time and being somewhat tetchy over my concerns. I really was worried about making some horrible mistake in restarting the `world` and all its Components by re-creating everything. Like ending up with two of everything and nothing getting rendered or only the ghost of the last game getting rendered.

I have relying on Klondike Tutorial Step 2, Scaffolding, which is well-explained but a little on the "monkey see, monkey do" side. I had no deep understanding of what I was doing, hence my worries. But you have shown me how to create my own World and be in full control of its initialization and re-initialization with various initial parameters. I would never have thought of that. I'll get right onto it tomorrow. As they say, you never learn something fully until you have to teach it... and a new World will be a great tutorial topic.

As to the square brackets, I had completely forgotten how to do optional positional parameters! I thought Random was wanting some kind of list!

Re passing values up from KlondikeGame and down to the next instance of it, I do it by defining a Dart data-only class at the top, containing a mix of input values, output values (e.g. result of game) and input/output values (e.g. klondikeDraw, type of startup and re-deal info). Then I pass a reference to an object of the data-only type to and fro. I used it in my pure-Flutter Sudoku game to pass data into and out of an Isolate and also from a menu screen to a game instance via the Router.

I also would like everything to be top notch and I have another motive: I would like to reproduce KDE's KPat game in Flame... see https://linuxreviews.org/KPatience - old, but the game itself is still going strong.

Reply:
Klondike with the new KlondikeWorld is working (roughly) now. The buttons and restart options are not properly connected yet, nor is the random number generator logic: but the gameplay and animations seem OK.

I was not happy about having to add a `KlondikeWorld world` parameter to the `Card` class, to get it to compile. The only `world` that is known at compile-time is the one in `FlameGame`. Maybe there is a more elegant way to handle this problem...

Reply:
> I was not happy about having to add a KlondikeWorld world parameter to the Card class, 

I suggested a solution for that here:
https://github.com/flame-engine/flame/pull/2842#discussion_r1390435457

Reply:
Fixed the Card class as you suggested. Also hooked KlondikeGame, KlondikeWorld, all the Components and KlondikeGame together (I hope). It works, but I don't think it is "top notch". It's been mis-compiling and crashing all over the landscape all day. So I would appreciate your expert eye being cast over the `klondike_game.dart` file especially. It contains the KlondikeGame and KlondikeWorld classes.

Reply:
How was it been crashing?